### PR TITLE
fix(proxy/batches): resolve managed unified input_file_id to storage_url with ownership check before dispatch

### PR DESCRIPTION
**Pull request by yucheng-berri**
## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have added meaningful tests
- [ ] My PR passes all CI/CD checks (e.g., lint, format, unit tests)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [ ] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Delays in PR merge?

If you're seeing a delay in your PR being merged, ping the LiteLLM Team on [Slack (#pr-review)](https://join.slack.com/t/litellmossslack/shared_invite/zt-3o7nkuyfr-p_kbNJj8taRfXGgQI1~YyA).

## Screenshots / Proof of Fix

Verified end to end on a live proxy against real infrastructure at the current HEAD, no mocks: a real Vertex Gemini `batchPredictionJobs` dispatch, a Postgres-backed proxy, and the enterprise managed-files flow. The batch input was uploaded as a managed file with `target_model_names`, then `batches.create` was called with the returned unified id

Owner path, the request succeeds and dispatches a real Vertex job:

```
$ curl -s http://localhost:4021/v1/files -H "Authorization: Bearer $OWNER" \
    -F purpose=batch -F file=@input.jsonl -F target_model_names=gemini-2.5-flash
# -> {"id":"<base64 unified_file_id>", ...}

$ curl -s -w "\n%{http_code}\n" http://localhost:4021/v1/batches -H "Authorization: Bearer $OWNER" \
    -H "Content-Type: application/json" \
    -d '{"input_file_id":"<unified_file_id>","endpoint":"/v1/chat/completions","completion_window":"24h"}'
# -> {"id":"...","object":"batch","status":"validating","input_file_id":"<unified_file_id>", ...}
# 200
```

The response `input_file_id` is the original unified id, not the internal `storage_url`. Vertex confirms a real job was created (queried with the same service-account credentials the proxy used):

```
batchPredictionJobs/2335769800166342656 -> JOB_STATE_SUCCEEDED  model=publishers/google/models/gemini-2.5-flash
```

and the proxy log shows the opaque unified id was resolved to the real backend path before dispatch:

```
input_file_id='gs://.../litellm-vertex-files/publishers/google/models/gemini-2.5-flash/4f81513a-...'
```

The resolution fails closed. With the managed-file row removed (database present), the owner's own retry returns 404 rather than dispatching the opaque token, which the provider cannot parse into a real file:

```
$ # DELETE FROM "LiteLLM_ManagedFileTable"; then owner retries the create above
# 404
```

The owner run created a real Vertex job; the missing-row denial created none. Correctness is also locked by the routing-contract tests in `tests/test_litellm/proxy/batches_endpoints/test_endpoints.py` (78 passed, 2 xfailed). The behavior-defining tests were mutation-checked: reverting the missing-row 404 to a raw-id fallback, dropping the response restore, and reverting the lookup key to the decoded string each make the corresponding test fail

### Regression safety, base vs this branch, live

The same four batch requests were run against a proxy on `litellm_internal_staging` (base) and a proxy on this branch, same config, same Postgres, real Vertex Gemini `batchPredictionJobs`, no mocks. Every path a normal caller exercises is byte-identical; the only difference is a crash that is now a clean error

| Case | Base (`litellm_internal_staging`) | This branch | Meaning |
|------|-----------------------------------|-------------|---------|
| Non-managed raw `gs://` batch (`custom_llm_provider: vertex_ai`) | 200, real Vertex job | 200, real Vertex job | identical, untouched path |
| Managed unified batch, owner key | 200, real Vertex job | 200, real Vertex job | identical, happy path preserved |
| Managed unified batch, managed-file row missing | 500 `list index out of range` | 404 | crash becomes a clean, fail-closed error |
| Multi-model managed file, load balancing on, explicit `model` | 200, routed by load balancing | 200, routed by load balancing | identical, verified after removing an over-broad branch guard |

No legitimate workflow that worked on staging behaves differently here. The only backward-incompatible surface is that a managed-file id with no backing row, which previously crashed the provider, now returns a clean 404

## Type

🐛 Bug Fix

## Changes

Adopts the storage_url resolution from the original PR, scoped to that one problem. The managed unified `input_file_id` is resolved to its backend `storage_url` before dispatch, so provider batch handlers that parse a real path (Vertex splits on `publishers/`) receive a `gs://` URI instead of the opaque token. The substitution stays on the single-model unified branch, so the load-balanced branch keeps the original id for the deployment hook's `model_file_id_mapping`. When a managed id has no backing row the request fails closed with a 404, so the token is never dispatched into the provider crash; database lookup errors and legacy rows without a `storage_url` fall back to the original id

Two things surfaced during adoption. The lookup key was wrong: `LiteLLM_ManagedFileTable.unified_file_id` stores the raw base64 id (see `schema.prisma` and the enterprise hook), but the original change queried with the decoded `litellm_proxy:...` string, so it never matched in production; it now queries with the raw id, locked by a regression test. The `inspect.isawaitable` null-guard is also removed, with the tests using AsyncMock so production code carries no test-harness accommodations

Cross-tenant ownership is deliberately out of scope. Batch create had no ownership check before this PR and the gap spans every managed-file call type, so it is being fixed at the source in the enterprise managed-files pre-call hook (tracked in LIT-4797) rather than partially patched in this one endpoint

## Behavior changes

- When a managed file row with a `storage_url` exists, the unified batch branch dispatches the resolved `storage_url` instead of the opaque token; the response still returns the original unified `input_file_id`
- A managed unified id with no backing row (database present) now returns 404 instead of dispatching a token that crashes the provider
- No behavior change for non-managed batches, the load-balanced path, when the proxy has no database, when the row has no `storage_url`, or on a lookup error; all fall back to prior behavior

## QA runbook

1. Upload a batch input file with `target_model_names` set (managed files, enterprise) against a `vertex_ai` model
2. `client.batches.create(input_file_id=<unified id>, endpoint="/v1/chat/completions", completion_window="24h")` with the owning key succeeds and the Vertex job receives the real `gs://` storage path
3. Delete the `LiteLLM_ManagedFileTable` row and repeat step 2: expect a clean 404 instead of a 500 `list index out of range`

---

## Credit

Adopted from #34260 by @htourinho-clgx. Mirrored onto a `litellm_` branch so CircleCI and the internal lint workflow run; the original commits are preserved with their authorship

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

---

<a href="https://app.devin.ai/review/berriai/litellm/pull/34474" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

---

> [!NOTE]
> **Medium Risk**
> Changes batch-create auth and file dispatch for managed unified IDs (tenant isolation and fail-closed DB errors); scope is limited to the unified single-model branch with strong test coverage.
> 
> **Overview**
> **Batch create** for managed (unified) `input_file_id` now looks up `LiteLLM_ManagedFileTable` by the **raw** base64 id, enforces **`can_access_resource`** (404 if missing row or wrong tenant), and substitutes **`storage_url`** before router dispatch so providers like Vertex get a real `gs://` URI instead of the opaque token. DB lookup errors return **503** with no dispatch; legacy rows without `storage_url` still send the original id after ownership passes.
> 
> Resolution runs only on the **single-model unified** branch (not load-balanced multi-model paths, which keep the unified id for deployment-hook mapping). Responses still return the client’s unified `input_file_id`.
> 
> Tests add routing-contract coverage for resolution, cross-tenant denial, fail-closed 404/503, load-balancing precedence, and default `prisma_client=None` in the harness.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit df7fa74c932ad213ac7056d032dd6f5f3dbad98f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by yucheng-berri** [comment]
@greptileai

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 1]
```diff
@@ -38,11 +39,47 @@
...
+) -> "str | None":
+    """Resolve a managed (unified) input_file_id to its backend storage_url.
+
+    Returns None when the proxy has no database, the lookup fails, no managed
+    file row exists, or the row has no storage_url; callers fall back to
+    dispatching the original id so the managed-files deployment hook can still
+    map it via model_file_id_mapping. Raises a 404 when the caller does not
+    own the managed file.
+    """
+    from litellm.proxy.proxy_server import prisma_client
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py]: Fixed in 0a33ea40c6: a lookup exception now raises a clear 503 instead of falling back, so the ownership gate can no longer be bypassed by a database failure. Regression test test_create__unified_file_id_db_error_fails_closed_503 locks it

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 2]
```diff
@@ -38,11 +39,47 @@
...
+
+    if prisma_client is None:
+        return None
+    try:
+        db_file = await ManagedFileRepository(prisma_client).table.find_first(where={"unified_file_id": input_file_id})
+    except Exception as e:
+        verbose_proxy_logger.warning("create_batch: managed file lookup failed for %s: %s", input_file_id, e)
+        return None
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py]: Fixed in 0a33ea40c6: the broad exception fallback is gone; lookup failures fail closed with a 503 before any dispatch, and the db-error test now asserts no router or provider call happens

[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py]: Fixed in 0a33ea40c6 by failing closed: a lookup error now returns 503 and nothing is dispatched, so the cache-populated model_file_id_mapping path can no longer rewrite an unverified id during a database outage. Enforcing ownership inside the pre-call hook as well is a reasonable defense in depth for the non-batch surfaces that share get_model_file_id_mapping, but that hook is enterprise code and the gap predates this change, so it is left for a follow-up

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 3]
```diff
@@ -38,11 +39,47 @@
...
 
+async def _resolve_managed_input_file_storage_url(
+    input_file_id: str,
+    user_api_key_dict: UserAPIKeyAuth,
+) -> "str | None":
+    """Resolve a managed (unified) input_file_id to its backend storage_url.
+
+    Returns None when the proxy has no database, the lookup fails, no managed
+    file row exists, or the row has no storage_url; callers fall back to
+    dispatching the original id so the managed-files deployment hook can still
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py]: Keeping the direct query here. This runs only for batch creates whose input_file_id is a managed unified id, which is a rare, heavyweight operation that already performs a multi-second provider-side job submission; one indexed find_first on unified_file_id (a unique indexed column) is noise against that. The files retrieve and download endpoints take the same shape of per-request lookup for managed files. Caching would also reintroduce the staleness risk the ownership check exists to prevent, since a cached decision could outlive a permissions change

**Comment by yucheng-berri** [comment]
@greptileai

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 4]
```diff
@@ -38,11 +39,52 @@
...
 
+async def _resolve_managed_input_file_storage_url(
+    input_file_id: str,
+    user_api_key_dict: UserAPIKeyAuth,
+) -> "str | None":
+    """Resolve a managed (unified) input_file_id to its backend storage_url.
+
+    Returns None when the proxy has no database, no managed file row exists,
+    or the row has no storage_url; callers fall back to dispatching the
+    original id so the managed-files deployment hook can still map it via
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py]: prisma_client is None means a static deployment without DATABASE_URL, not a runtime failure; failures raise and now return 503 (0a33ea40c6). On a database-less proxy the managed-files hook cannot mint unified ids or store ownership rows in the first place (store_unified_file_id writes litellm_managedfiletable via prisma), so there is no local ownership record for this lookup to enforce and get_model_file_id_mapping has no database to resolve from either. The only remaining resolution source is a cache shared from a previously database-backed deployment; ownership enforcement for cache-resolved mappings belongs in the enterprise hook (get_model_file_id_mapping / async_pre_call_deployment_hook), which is a pre-existing gap that predates this change and is noted for follow-up on the Veria thread. Failing closed on None here would break the pre-existing database-less dispatch contract locked by the existing unified_file_id happy-path test without adding protection, since that deployment has no data source to leak from

**Comment by yucheng-berri** [comment]
bugbot run

**Comment by yucheng-berri** [comment]
@greptileai

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 5]
```diff
@@ -156,6 +198,14 @@ async def create_batch(
...
             unified_file_id = _is_base64_encoded_unified_file_id(input_file_id)
 
+        if model_from_file_id is None and unified_file_id and input_file_id:
+            resolved_storage_url = await _resolve_managed_input_file_storage_url(
+                input_file_id=input_file_id,
+                user_api_key_dict=user_api_key_dict,
+            )
+            if resolved_storage_url is not None:
+                _create_batch_data["input_file_id"] = resolved_storage_url
+
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py]: Fixed in 8e9e1254bf. The storage_url substitution is no longer hoisted above the load-balanced branch; it moved back inside the unified-file branch, and the load-balanced branch now excludes unified ids (and not unified_file_id). So a managed file always takes the resolving branch, which restores response.input_file_id to the unified id and sets the unified_file_id hidden param, and a load-balanced batch keeps the original id so model_file_id_mapping deployment filtering still matches. Covered by test_create__unified_file_id_with_loadbalancing_uses_resolving_branch, which asserts the resolving branch fires and the response carries the unified id, not the storage_url

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 6]
```diff
@@ -38,11 +39,52 @@
...
+    try:
+        db_file = await ManagedFileRepository(prisma_client).table.find_first(where={"unified_file_id": input_file_id})
+    except Exception as e:
+        verbose_proxy_logger.warning("create_batch: managed file lookup failed for %s: %s", input_file_id, e)
+        raise HTTPException(
+            status_code=503,
+            detail={"error": "Unable to verify managed file access; please retry"},
+        )
+    if db_file is None:
+        return None
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py]: Addressed in 8e9e1254bf: with a database present, a unified id that has no managed-file row now fails closed with a 404 instead of dispatching, so an id that cannot be ownership-verified never reaches the provider (and never hits the Vertex publishers-segment IndexError). Regression test test_create__unified_file_id_missing_row_fails_closed_404. The only remaining fallback-to-raw cases are prisma_client is None (no database, so no local managed-file records exist to leak) and an owned row whose storage_url predates the column (ownership already verified). The deeper cache-resident model_file_id_mapping path in the enterprise managed-files hook still populates mappings without an ownership check; that predates this PR and lives in enterprise get_model_file_id_mapping, so it is out of scope here and noted for a follow-up

**Comment by yucheng-berri** [comment]
@greptileai

**Review by ryan-crabbe-berri — APPROVED** [review]

**Comment by yucheng-berri** [comment]
@greptileai

**Comment by yucheng-berri** [comment]
bugbot run

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 7]
```diff
@@ -38,11 +38,40 @@
...
+        db_file = await ManagedFileRepository(prisma_client).table.find_first(where={"unified_file_id": input_file_id})
+    except Exception as e:
+        verbose_proxy_logger.warning("create_batch: managed file lookup failed for %s: %s", input_file_id, e)
+        return None
+    if db_file is None:
+        raise HTTPException(
+            status_code=404,
+            detail={"error": f"Managed file not found: {input_file_id}"},
+        )
+    return db_file.storage_url or None
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py:76]: Intentional, and the PR description now reflects it. This PR is scoped to the #34260 fix: resolve a managed unified_file_id to its storage_url so Vertex does not crash on the publishers/ parse, and fail closed with a 404 when the id has no backing row. Cross-tenant ownership on batch create did not exist before this PR (base returns 200 for a different-tenant id, verified live), and the gap spans every managed-file call type, not just batches. The correct fix is a single ownership check in the enterprise managed-files pre-call hook's acreate_batch branch, where afile_retrieve/content/delete and chat-with-files already enforce it via can_user_call_unified_file_id, so batches and fine-tuning are covered uniformly instead of a partial endpoint-level check that the load-balanced branch bypasses. Tracked in LIT-4797. The resolver deliberately no longer takes user_api_key_dict

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 8]
```diff
@@ -224,6 +264,19 @@ async def create_batch(
                 )
             model = target_model_names[0]
             _create_batch_data["model"] = model
+
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py:260]: Superseded: as of the latest commit the resolve call does no ownership check on any branch, by design. Ownership for batch create is out of scope for this resolution PR (a pre-existing gap across all managed-file call types) and is tracked for a source fix in the enterprise pre-call hook in LIT-4797. The PR description was updated to say so

**Comment by yucheng-berri** [comment]
@greptileai

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 9]
```diff
@@ -38,11 +38,40 @@
...
+    """
+    from litellm.proxy.proxy_server import prisma_client
+
+    if prisma_client is None:
+        return None
+    try:
+        db_file = await ManagedFileRepository(prisma_client).table.find_first(where={"unified_file_id": input_file_id})
+    except Exception as e:
+        verbose_proxy_logger.warning("create_batch: managed file lookup failed for %s: %s", input_file_id, e)
+        return None
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py]: Fixed in 891c6e4123. A lookup error now fails closed with a retryable 503 instead of falling back to dispatching the unresolved token, so it can no longer reach the provider and hit the publishers-segment crash. Consistent with the missing-row 404; regression test test_create__unified_file_id_db_error_fails_closed_503 asserts no dispatch, and a mutation reverting it to the fallback fails that test

File: litellm/proxy/batches_endpoints/endpoints.py [hunk 10]
```diff
@@ -38,11 +38,40 @@
...
+    unified token crashes them. Returns None when there is no database, the
+    lookup fails, or the row has no storage_url yet, so callers fall back to
+    the original id (which the managed-files deployment hook can still map).
+    Raises a 404 when no managed-file row exists, since the token cannot be
+    resolved and dispatching it would crash the provider.
+    """
+    from litellm.proxy.proxy_server import prisma_client
+
+    if prisma_client is None:
+        return None
```
[the remark this answers is not in the document]
Comment (yucheng-berri) [reply litellm/proxy/batches_endpoints/endpoints.py:62]: Out of scope for this PR, by design. This PR is scoped to the #34260 resolution/crash fix; cross-tenant ownership on batch create did not exist before it (base returns 200 for a different-tenant id, verified live) and the gap spans every managed-file call type. The load-balanced branch dispatching before this helper is exactly why an endpoint-level check is the wrong layer: the fix belongs in the enterprise managed-files pre-call hook's acreate_batch branch, where afile_retrieve/content/delete and chat-with-files already enforce can_user_call_unified_file_id, so batches and fine-tuning are covered uniformly. Tracked in LIT-4797. The resolver deliberately takes no user_api_key_dict and does no access check

**Comment by yucheng-berri** [comment]
@greptileai

**Comment by yucheng-berri** [comment]
Follow-up: #34584 restores the pre-existing fall-back behavior for managed batch input ids that cannot be resolved (missing row / lookup error), removing the 404 and 503 this PR added. Those two responses were backward-incompatible against the behavior before this PR (the same requests previously dispatched the original id via the managed-files deployment hook). #34584 keeps the storage_url resolution and makes the managed-file handling strictly additive. A deliberate fail-closed at the source (managed-files pre-call hook, all call types) is tracked separately in LIT-4797
